aws-java-sdk-sts/src/main/java/com/amazonaws/auth/STSSessionCredentialsProvider.java

```diff
@@ -116,39 +167,186 @@ public STSSessionCredentialsProvider(AWSCredentialsProvider longLivedCredentials
+     * Unconditionally calls STS to generate a new session
+     * 
+     * @return 
      */
-    private void startSession() {
+    private SessionCredentialsHolder newSession() {
         GetSessionTokenResult sessionTokenResult = securityTokenService
-                .getSessionToken(new GetSessionTokenRequest().withDurationSeconds(DEFAULT_DURATION_SECONDS));
+                .getSessionToken(new GetSessionTokenRequest()
+                        .withDurationSeconds(DEFAULT_DURATION_SECONDS));
+
         Credentials stsCredentials = sessionTokenResult.getCredentials();
 
-        sessionCredentials = new BasicSessionCredentials(
+        AWSSessionCredentials credentials = new BasicSessionCredentials(
                 stsCredentials.getAccessKeyId(),
                 stsCredentials.getSecretAccessKey(),
                 stsCredentials.getSessionToken());
-        sessionCredentialsExpiration = stsCredentials.getExpiration();
+
+        return new SessionCredentialsHolder(credentials, stsCredentials.getExpiration());
+    }
+    
+    /**
+     * Used when there is no valid session to return. Callers are blocked until
+     * a new session is created or an exception is thrown.
+     * Throws AmazonClientException if new session is not created
+     * within 5 seconds.
+     */
+    private void blockingNewSession() {
+        newSessionLock.lock();
+
+        try {
+            // return if successful refresh occured while waiting for the lock
+            if(!expired()) {
+                return;
+            }
+            
+            // trigger a refresh thread if needed
+            asyncNewSession();
+
+            // wait for new session or time out
+            if (false == newSessionCondition.await(NEW_SESSION_MAX_WAIT_SECONDS, TimeUnit.SECONDS)) {
+                // timeout occured, check for valid session before throwing an exception
+                if (expired()) {
+                    throw new AmazonClientException("Session credentials request timed-out.", lastException.get());
+                }
+            }
+        } catch (InterruptedException ex) {
+            throw new AmazonClientException(ex);
+        } finally {
+            newSessionLock.unlock();
+        }
+    }
+    
+    /**
+     * Used to asynchronously refresh the session. Ensures only one
+     * thread is spawned at any given time to refresh the session. Caller
+     * is never blocked.
+     */
+    private void asyncNewSession() {
+        // immediate return if refresh already in progress
+        if (refreshing.compareAndSet(false, true)) {
+            try {
+                if (false == asyncRefreshing.get()) {
+                    new Thread() {
+                        @Override
+                        public void run() {
+                            // thread exits immediately if an async refresh already in progress
+                            if (asyncRefreshing.compareAndSet(false, true)) {
```

Comment:
My concern here is what happens if ```new``` or ```Thread.start()``` throws an exception. I prefer that to be at least potentially recoverable. If the campareAndSet is moved out of the thread, it would require intricate exception handling with maintainability issues. However, since ```refreshing``` is no longer being used external to ```asyncNewSession```, I think this logic can be simplified.